### DISCUSSION TITLE
New LAT Tracking Issue

### DISCUSSION THREAD

**Discussion by skhrg** [simonsobs/sodetlib#423] [Issues and Debugging] [answered]
LAT UFMs aren't tracking at all anymore. We see this issue across multiple bands and slots. Below are plots from crate 1, slot 2, band 0.

Tuning goes fine:
![image](https://github.com/simonsobs/sodetlib/assets/13905516/fd85df54-7a68-4aa7-9704-0cf619949414)
 
And the resonators are looking good:
![image](https://github.com/simonsobs/sodetlib/assets/13905516/32f127ff-3164-413a-87a5-e45344429b59)

But when we try to track we get 0 dets:
![image](https://github.com/simonsobs/sodetlib/assets/13905516/2f035f1c-f67a-46d7-a723-2a854c9f716d)

Looking at the tracking plots its garbage like this:
![image](https://github.com/simonsobs/sodetlib/assets/13905516/e6ed88b1-763f-44eb-b31b-0fde56abfc9f)

With the squid curve like this:
![image](https://github.com/simonsobs/sodetlib/assets/13905516/af3362bd-ba86-435b-b1d1-308c933a0726)

Turning off feedback things look like this:
![image](https://github.com/simonsobs/sodetlib/assets/13905516/1dc617b8-e24d-4b0f-939d-5ee1e151dded)
![image](https://github.com/simonsobs/sodetlib/assets/13905516/19901c76-5332-4c7d-900a-72490db98ec3)

The original theory was that the ADCs were saturating (see older tracking plot below):
![image](https://github.com/simonsobs/sodetlib/assets/13905516/6c6b9f1f-f224-4459-923e-071e293c22b1)
but all of the above plots were with the attens at 30 and not close to saturating.

The magic hammer technique also doesn't make any difference.

**Comment by swh76** [comment, marked as answer]
@skhrg and I looked at this and found the fix.  This was caused by a subtle bug in pysmurf `setup` here - https://github.com/slaclab/pysmurf/blob/e8db8010a6698beedae4e55928ab858f6c7cd0d6/python/pysmurf/client/base/smurf_control.py#L551.

All of the plots on ticket look nominal (except the final tracking plot), but while looking at rogue logs (via `docker logs smurf-streamer-s2` on smurf-srv-SO6);

```
...
ERROR:pyrogue.Variable.RemoteVariable.AMCc.FpgaTopLevel.AppTop.AppCore.SysgenCryo.Base[0].lmsDelay:"Memory Error for AMCc.FpgaTopLevel.AppTop.AppCore.SysgenCryo.Base[0].lmsDelay at address 0x8100010c Invalid value '88' for base type <class 'int'> with bit size 6"
ERROR:pyrogue.Variable.RemoteVariable.AMCc.FpgaTopLevel.AppTop.AppCore.SysgenCryo.Base[0].lmsDelay:Error setting value '88' to variable 'AMCc.FpgaTopLevel.AppTop.AppCore.SysgenCryo.Base[0].lmsDelay' with type UInt6. Exception="Memory Error for AMCc.FpgaTopLevel.AppTop.AppCore.SysgenCryo.Base[0].lmsDelay at address 0x8100010c Invalid value '88' for base type <class 'int'> with bit size 6"
1710797843.796532:pyrogue.epicsV3.Value: Error setting value from epics: smurf_server_s2:AMCc:FpgaTopLevel:AppTop:AppCore:SysgenCryo:Base[0]:lmsDelay
...
```

that's the error for band 0 (which corresponds to `Base[0]` in the register path) but there's a similar error for all 8 bands.  The bug is that factor of 4 in https://github.com/slaclab/pysmurf/blob/e8db8010a6698beedae4e55928ab858f6c7cd0d6/python/pysmurf/client/base/smurf_control.py#L551 which is overflowing the lmsDelay register ; the history is that the SMuRF firmware used to process four tones in every polyphase filter bank (PFB) subband, but the production firmware we use now processes only one tone in every PFB subband.  The 4* was only relevant for older versions of the firmware where there was more delay due to each subband needing to do four times more processing.

I will open an issue on pysmurf to fix this, and leave this ticket open until it's fixed.  The reason other users haven't noticed is that `lmsDelay` is properly set by the `S.estimate_phase_delay` function (and its variants), which triggers the low level rogue `setBandDelay` routine (https://github.com/slaclab/cryo-det/blob/797fa80aa0e43bc4c0939921702a053f3f69ed2b/firmware/python/CryoDet/DspCoreLib/CryoDetCmbHcd/_CryoFreqBand.py#L661).  Here's that function for reference if you don't have access to the `cryo-det` repo;

```
def setBandDelay(dev, var, value, write):
        # workaround for rogue local variables
        # list objects get written as string, not list of float when set by GUI
        if isinstance(value, str):
            value = eval(value)

        refPhaseDelaySteps = 2.4       # 2.4MHz steps
        refPhaseDelayFineSteps = 307.2 # adds 307.2MHz steps
        # ceil makes us overcorrect
        refPhaseDelay = int(np.ceil(value*refPhaseDelaySteps))
        # lmsDelay is in 2.4MHz ticks
        lmsDelay      = int(refPhaseDelay)
        # now add delay to compensate for overcorrection
        overCorrect = refPhaseDelay-value*refPhaseDelaySteps
        refPhaseDelayFine = int(np.round((refPhaseDelayFineSteps/refPhaseDelaySteps) * overCorrect))
        if refPhaseDelay > 39:
            print(f'Warning cacluated refPhaseDelay is {refPhaseDelay}, max allowable is 39')
            refPhaseDelay = int(39)
            refPhaseDelayFine = int(0)
            lmsDelay = int(39)

        var.dependencies[0].set( refPhaseDelay, write=write )     # refPhaseDelay
        var.dependencies[1].set( refPhaseDelayFine, write=write ) # refPhaseDelayFine
        var.dependencies[2].set( lmsDelay, write=write )     # lmsDelay
```

the last line of the function properly sets the lmsDelay register, overwriting the bug in `setup`.  Here's band 0 on slot 2 of the SMuRF crate connected to smurf-srv-SO6 properly tracking after running `estimate_phase_delay` on that band;

![Screenshot 2024-03-25 at 2 30 42 PM](https://github.com/simonsobs/sodetlib/assets/20823858/1e02e202-8c78-421b-aec0-63bb2369e1e8)

Users should be able to rely on the pysmurf cfg to set this ; will keep this issue open here until I release a fix in pysmurf.

**Comment by swh76** [comment]
Opened issue on pysmurf for this here - https://github.com/slaclab/pysmurf/issues/789.

**Comment by jlashner** [comment]
Thanks for tracking this down shawn.

Talked to saianeesh this morning and wanted to add some more context on the sodetlib side... so the reason why we missed this is because the pysmurf-config file was updated to have the correct phase delays. This means that the band_phase_delays were being set correctly by S.setup, while the lms_delay was not being set correctly because of the bug shawn found, which is why the tune looks fine while the tracking looks bad.

Normally, we don't usually update the values in the pysmurf-config file, so missing the `setup_phase_delay` step would've been caught, since the tune would've been bad. Good to know that this is also a secondary failure mode of forgetting estimate_phase_delay.

**Comment by swh76** [comment]
@skhrg @jlashner merged in pysmurf PR https://github.com/slaclab/pysmurf/pull/790.
